sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/util/ReduceFnRunner.java

```diff
@@ -326,6 +326,15 @@ private void collectAndMergeWindows(Iterable<WindowedValue<InputT>> values) thro
         @SuppressWarnings("unchecked")
         W window = (W) untypedWindow;
 
+        ReduceFn<K, InputT, OutputT, W>.Context directContext =
+            contextFactory.base(window, StateStyle.DIRECT);
+        if (triggerRunner.isClosed(directContext.state())) {
```

Comment:
Does this mean that we may not have ensured it exists? I guess the logic is if we never remove the window from the active window set, then it should already exist, right?